source/credentials_provider_sts_web_identity.c

```diff
@@ -0,0 +1,1120 @@
+    }
+
+    /* As long as we parsed an usable expiration, use it, otherwise use
+     * the existing one: now + 900s, initialized before parsing.
+     */
+    if (aws_byte_cursor_eq_c_str_ignore_case(&node->name, "Expiration")) {
+        aws_xml_node_as_body(parser, node, &credential_data);
+        if (credential_data.len != 0) {
+            struct aws_date_time expiration;
+            if (aws_date_time_init_from_str_cursor(&expiration, &credential_data, AWS_DATE_FORMAT_ISO_8601) ==
+                AWS_OP_SUCCESS) {
+                credentials->expiration_timepoint_seconds = (uint64_t)aws_date_time_as_epoch_secs(&expiration);
+            }
+        }
+    }
+    return true;
+}
+
+static struct aws_credentials *s_sts_web_identity_parse_credentials_from_response(
+    struct aws_allocator *allocator,
+    struct aws_byte_buf *response) {
+
+    if (!allocator || !response || response->len == 0) {
+        return NULL;
+    }
+
+    struct aws_xml_parser xml_parser;
+    struct aws_byte_cursor response_cursor = aws_byte_cursor_from_buf(response);
+    if (aws_xml_parser_init(&xml_parser, allocator, &response_cursor, 0)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to init xml parser for sts web identity credentials provider to parse error information.")
+        return NULL;
+    }
+    uint64_t now = UINT64_MAX;
+    if (aws_sys_clock_get_ticks(&now) != AWS_OP_SUCCESS) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to get sys clock for sts web identity credentials provider to parse error information.")
+        return NULL;
+    }
+    uint64_t now_seconds = aws_timestamp_convert(now, AWS_TIMESTAMP_NANOS, AWS_TIMESTAMP_SECS, NULL);
+    struct aws_credentials *credentials = aws_mem_calloc(allocator, 1, sizeof(struct aws_credentials));
+    if (!credentials) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to allocate required memory for credentials: %s",
+            aws_error_str(aws_last_error()));
+        return NULL;
+    }
+    credentials->allocator = allocator;
+    credentials->expiration_timepoint_seconds = now_seconds + STS_WEB_IDENTITY_CREDS_DEFAULT_DURATION_SECONDS;
+
+    if (aws_xml_parser_parse(&xml_parser, s_on_creds_node_encountered_fn, credentials)) {
+        AWS_LOGF_ERROR(
+            AWS_LS_AUTH_CREDENTIALS_PROVIDER,
+            "Failed to parse xml response for sts web identity with error: %s",
+            aws_error_str(aws_last_error()));
+        aws_xml_parser_clean_up(&xml_parser);
+        aws_credentials_destroy(credentials);
+        return NULL;
+    }
+
+    aws_xml_parser_clean_up(&xml_parser);
+    if (!credentials->access_key_id || !credentials->secret_access_key) {
+        aws_credentials_destroy(credentials);
+        return NULL;
+    }
+    return credentials;
```

Comment:
Given all the error early outs, this function would be better with success path and an error label that cleaned up everything (creds, parser)